src/com/google/javascript/jscomp/GlobalNamespace.java

```diff
@@ -1123,7 +1149,7 @@ boolean shouldKeepKeys() {
     }
 
     boolean needsToBeStubbed() {
-      return globalSets == 0 && localSets > 0;
+      return globalSets == 0 && localSets > 0 && localSetsWithNoCollapse <= 0;
```

Comment:
I think <code>@nocollapse</code> should only have an effect on the declaration site of a name, not on any random assignment to the name.